### PR TITLE
feat(data-events): donation listeners

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR tweaks the hook into `woocommerce_checkout_order_processed` so it's more easily accessible for other parts of the code that might want to reach donation processing. There shouldn't be any change into how the `sync_reader_from_order()` behaves.

The following donation-related listeners are added:

### `donation_new`

When there's a new donation, either through Stripe or Newspack (WooCommerce) platforms.

### Data

| Name            | Type     |
| --------------- | -------- |
| `user_id`       | `int`    |
| `email`         | `string` |
| `amount`        | `float`  |
| `currency`      | `string` |
| `recurrence`    | `string` |
| `platform`      | `string` |
| `platform_data` | `array`  |

### `donation_subscription_new`

When there's a new WooCommerce Subscription. This action does not replace the `donation_new` that create the subscription.

| Name            | Type     |
| --------------- | -------- |
| `user_id`       | `int`    |
| `email`         | `string` |
| `amount`        | `float`  |
| `currency`      | `string` |
| `recurrence`    | `string` |
| `platform`      | `string` |
| `platform_data` | `array`  |

### `donation_subscription_cancelled`

When a WooCommerce Subscription is cancelled.

| Name              | Type     |
| ----------------- | -------- |
| `subscription_id` | `int`    |
| `user_id`         | `int`    |
| `email`           | `string` |
| `amount`          | `float`  |
| `currency`        | `string` |
| `recurrence`      | `string` |
| `platform`        | `string` |

### `donation_subscription_changed`

When a WooCommerce Subscription status changes.

| Name              | Type     |
| ----------------- | -------- |
| `subscription_id` | `int`    |
| `user_id`         | `int`    |
| `email`           | `string` |
| `status_before`   | `string` |
| `status_after`    | `string` |
| `amount`          | `float`  |
| `currency`        | `string` |
| `recurrence`      | `string` |
| `platform`        | `string` |

### How to test the changes in this Pull Request:

1. Make sure you have the following constants:
```
define( 'NEWSPACK_AMP_PLUS_ENABLED', true );
define( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION', true );
define( 'NEWSPACK_LOG_LEVEL', 2 );
```
2. Make sure you have the Reader Revenue Newspack (WooCommerce) platform configured
3. Make a new one-time donation using the donation block
4. Confirm you see the logs with `[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "donation_new" via Action Scheduler.`
5. Visit WooCommerce's Status -> Scheduled Actions
6. Confirm you see the `newspack_data_events_as_dispatch ` hook with `donation_new` as well
7. Now make a recurring donation (subscription) and confirm you see the same with: `donation_subscription_new`, `donation_subscription_changed` (switching from `pending` to `active`)
8. WIth Reader Activation enabled, make sure you have ActiveCampaign configured with a master list
9. Make a donation with a new email and confirm the contact data with donation details is being updated to ActiveCampaign as expected
10. Switch the platform to Stripe and repeats steps from 3 to 7

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/data-events/listeners.php
Comment (dkoo) [inline includes/data-events/listeners.php:161]: Testing these steps:

> 3. Make a new one-time donation using the donation block
> 4. Confirm you see the logs with [NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "donation_new" via Action Scheduler.

Making a one-time donation, I see both `donation_subscription_new` and `donation_new` actions being dispatched.

```
[12-Jan-2023 23:31:07 UTC] [NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "donation_subscription_new" via Action Scheduler.
[12-Jan-2023 23:31:07 UTC] [NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "donation_new" via Action Scheduler.
```

I also see this in the Scheduled Actions log:

Screen Shot 2023-01-12 at 5 17 26 PM

Note also the empty timestamps of the scheduled actions—not sure if that's a bug or not?

File: includes/data-events/listeners.php
Comment (dkoo) [inline includes/data-events/listeners.php:199]: Testing this step:

> 7. Now make a recurring donation (subscription) and confirm you see the same with: donation_subscription_new, donation_subscription_changed (switching from pending to active)

The `donation_subscription_changed` action is dispatched going from `active` to `pending` instead of `pending` to `active` as described above:

Screen Shot 2023-01-12 at 5 22 45 PM

Again, not sure if it's a bug or not, but it does differ from the testing steps.

Comment (dkoo) [reply includes/data-events/listeners.php:199]: It looks like the order of arguments here is off. According to [WC's docs](https://woocommerce.com/document/subscriptions/develop/action-reference/#section-2) for the `woocommerce_subscription_status_updated` action hook, the second argument is the "new" status (`status_to`) and the third is the old (`status_from`).

**Review by dkoo — CHANGES_REQUESTED** [review]
Ran into one issue where the `donation_subscription_new` action is getting dispatched upon a new single donation order, and a couple of questions which may or may not be issues.

File: includes/data-events/listeners.php
Comment (leogermani) [inline includes/data-events/listeners.php:176]: Thinking from an analytics POV, it would make sense to also have the subscription details here: amount, currency, recurrence and platform.

In case we are triggering non identifiable events and analyzing them, we could then compare to see what kind of subscriptions are being cancelled.

Comment (leogermani) [reply includes/data-events/listeners.php:176]: same for status_update.

makes sense?

File: includes/data-events/listeners.php
Comment (leogermani) [inline includes/data-events/listeners.php:164]: Nice! That's a great use of listeners I haven't thought about... same hook but dispatching different actions depending on the params

File: includes/data-events/README.md
Comment (miguelpeixe) [reply includes/data-events/README.md:77]: Good call! bb409c13105b650e49d8917bfaf1a74f3e9ad051 removes this heading from all instances

File: includes/data-events/listeners.php
Comment (miguelpeixe) [reply includes/data-events/listeners.php:216]: Indeed! Fixed with 8671e7b6b5654da06a17bc4bb5f24c6bdf3ee0db

File: includes/data-events/listeners.php
Comment (miguelpeixe) [reply includes/data-events/listeners.php]: Good call! 47e503963a39cd19212588cdee5a6810cb9318e6 expands the data for both dispatches

**Comment by miguelpeixe** [comment]
@dkoo, were you able to confirm that the AC integration works as expected while using WooCommerce? It worked without issues for me:

image

File: includes/data-events/listeners.php
Comment (dkoo) [reply includes/data-events/listeners.php:166]: Can confirm that the `donation_subscription_new` is only fired on new subscriptions now, not one-time donations.

The timestamps still look like `0000-00-00`, but I'm not sure if this is a bug.

File: includes/data-events/listeners.php
Comment (dkoo) [reply includes/data-events/listeners.php:216]: Confirmed!

**Review by dkoo — APPROVED** [review]
The updates look good to me, and I note the additional data on the events based on @leogermani's suggestion. I'd wait until @leogermani gives approval too before merging, though.

I did still have one question about the timestamps of scheduled actions looking like `0000-00-00 00:00:00`, but I'm not sure if this is an issue because the proper timestamp is logged in the "Log" column. So, I'll approve this at your and @leogermani's discretion.

**Comment by miguelpeixe** [comment]
Thank you, @dkoo! I noticed the `0000-00-00 00:00:00` timestamp as well. It looks like the default behavior when enqueueing a new task without scheduling, so it executes as soon as possible.

Taking a look at the AS repo, I noticed [this recently merged PR](https://github.com/woocommerce/action-scheduler/pull/898). Looks like it changes this default behavior also to improve the scheduling so overdue tasks are processed before the enqueued ones.

**Review by leogermani — APPROVED** [review]
This looks good to me.

In the future I'll check if it's possible to add some more data to the events to be used in the analytics, like where the donation came from? (from a prompt, sdb, etc...)

**Pull request merged**
